p2p/peer.py

```diff
@@ -143,6 +143,106 @@
     return peer
 
 
+class HeaderRequest(NamedTuple):
+    block_number_or_hash: BlockIdentifier
+    max_headers: int
+    skip: int
+    reverse: bool
+
+    def generate_block_numbers(self,
+                               block_number: BlockNumber=None) -> Tuple[BlockNumber, ...]:
+        if block_number is None and not self.is_numbered:
```

Comment:
Ok, now it's clear why we sometimes have to pass the block number here. Could we change `HeaderRequest` to have a `block_number: int` attribute and force callsites to to the header lookup before instantiating a `HeaderRequest`, when the peer specifies a block hash?

I'd really like to address this somehow because it's very easy to misuse the current API and get a `TypeError` when calling `generate_block_numbers()`